protocols/request-response/src/json.rs

```diff
@@ -0,0 +1,195 @@
+    use std::{io, marker::PhantomData};
+
+    /// Max request size in bytes
+    const REQUEST_SIZE_MAXIMUM: usize = 1024 * 1024;
+    /// Max response size in bytes
+    const RESPONSE_SIZE_MAXIMUM: usize = 10 * 1024 * 1024;
+
+    pub struct Codec<Req, Resp> {
+        phantom: PhantomData<(Req, Resp)>,
+    }
+
+    impl<Req, Resp> Default for Codec<Req, Resp> {
+        fn default() -> Self {
+            Codec {
+                phantom: PhantomData,
+            }
+        }
+    }
+
+    impl<Req, Resp> Clone for Codec<Req, Resp> {
+        fn clone(&self) -> Self {
+            Self::default()
+        }
+    }
+
+    #[async_trait]
+    impl<Req, Resp> crate::Codec for Codec<Req, Resp>
+    where
+        Req: Send + Serialize + DeserializeOwned,
+        Resp: Send + Serialize + DeserializeOwned,
+    {
+        type Protocol = StreamProtocol;
+        type Request = Req;
+        type Response = Resp;
+
+        async fn read_request<T>(&mut self, _: &Self::Protocol, io: &mut T) -> io::Result<Req>
+        where
+            T: AsyncRead + Unpin + Send,
+        {
+            let vec = read_to_end(io, REQUEST_SIZE_MAXIMUM).await?;
+
+            Ok(serde_json::from_slice(vec.as_slice())?)
+        }
+
+        async fn read_response<T>(&mut self, _: &Self::Protocol, io: &mut T) -> io::Result<Resp>
+        where
+            T: AsyncRead + Unpin + Send,
+        {
+            let vec = read_to_end(io, RESPONSE_SIZE_MAXIMUM).await?;
+
+            Ok(serde_json::from_slice(vec.as_slice())?)
+        }
+
+        async fn write_request<T>(
+            &mut self,
+            _: &Self::Protocol,
+            io: &mut T,
+            req: Self::Request,
+        ) -> io::Result<()>
+        where
+            T: AsyncWrite + Unpin + Send,
+        {
+            let data = serde_json::to_vec(&req)?;
+
+            io.write_all(data.as_ref()).await?;
+            io.flush().await?;
+            io.close().await?;
+
+            Ok(())
+        }
+
+        async fn write_response<T>(
+            &mut self,
+            _: &Self::Protocol,
+            io: &mut T,
+            resp: Self::Response,
+        ) -> io::Result<()>
+        where
+            T: AsyncWrite + Unpin + Send,
+        {
+            let data = serde_json::to_vec(&resp)?;
+            io.write_all(data.as_ref()).await?;
+            io.flush().await?;
+            io.close().await?;
+
+            Ok(())
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::Codec;
+    use futures_ringbuf::Endpoint;
+    use libp2p_swarm::StreamProtocol;
+    use serde::{Deserialize, Serialize};
+
+    #[async_std::test]
+    async fn test_codec() {
+        let expected_request = TestRequest {
+            payload: "test_payload".to_string(),
+        };
+        let expected_response = TestResponse {
+            payload: "test_payload".to_string(),
+        };
+        let protocol = StreamProtocol::new("/test_cbor/1");
```

Comment:
```suggestion
        let protocol = StreamProtocol::new("/test_json/1");
```

Not that it makes a functional difference but we are in the JSON module here.